### DISCUSSION TITLE
newbie question

### DISCUSSION THREAD

**Discussion by ravi567** [community/community#22878] [New to GitHub]
<p>I have been using git for few months. I thought I had understood the behaviour of file changes within a branch in a certain way. I am seeing some different behavior. Could someone help clarify.</p>
<p>I have 2 branches master and dev.</p>
<ol>
<li>
<p>I create a new file called test_file in dev branch. When I run git status, it shows me test_file is a untracked file now.</p>
</li>
<li>
<p>I run a git checkout master to switch to the master branch. When I run a git status, it shows me the test_file is a untracked file there as well.</p>
</li>
</ol>
<p>Question: My assumption was a file created in one branch will not be seen in another branch, unless I have commited the branch (dev) and merged it with other branch (master). Is my understanding wrong?</p>
<ol>
<li>If I add the test_file to index (git add test_file), commit the file in devl branch, I no longer see the file in master branch.</li>
</ol>
<p>Question: does this mean the file will not be seen in other branches (master) only when it’s commited in a branch (dev).</p>
<p>Thanks</p>
<p>Ravi</p>

**Comment by dtashley** [comment, marked as answer]
<p>If I’m understanding your description correctly, the issue was likely that test_file was untracked, so git let it hang around in your sandbox while you switched branches.</p>
<p>I have not verified this behavior, but it makes sense if this is the case.</p>
<p>When you switch branches in your git, git’s only rule is not to destroy information.  This means that if a tracked file has been modified, it should refuse to do the switch.  However, if a tracked file has not been modified, it is free to bring a different revision into your sandbox or to delete it from your sandbox (the rationale for this is that you can get the file and its contents back by switching back to the original branch).</p>
<p>But with an untracked file, git’s designers likely had to decide between two options:</p>
<p>1)Refuse to do the switch until the user has disposed of the untracked file.</p>
<p>2)Do the switch, but leave the untracked file in place, unchanged.</p>
<p>Note that the following would NOT have been an option:</p>
<p>3)Delete the tracked file.  (Not an option because you could not recover the file.)</p>
<p>I’m guessing that git’s designers chose (2) because it is less obtrusive.  It would be a common use case that developers have extra untracked files in the sandbox when switching branches.</p>
<p>(*) One case that I would be curious about is if you are on a branch where test_file is not tracked, you create test_file, then try to switch to a branch where test_file IS tracked.  I’m betting git will refuse to let you switch branches.</p>
<p>Again, git’s priority would be to not destroy your data.  Leaving an untracked file in your sandbox while switching branches does not destroy data.</p>
<p>If you test (*), let me know the results.</p>

**Comment by sfynx** [comment]
<blockquote>
<p><a class="mention" href="/dtashley">@dtashley</a> wrote:</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>(*) One case that I would be curious about is if you are on a branch where test_file is not tracked, you create test_file, then try to switch to a branch where test_file IS tracked.  I’m betting git will refuse to let you switch branches.</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>Correct, it is blocked, stating the files that would have been overwritten so you know what to get to safety first.</p>
